sql/log_event_server.cc

```diff
@@ -6068,6 +6075,9 @@ int Rows_log_event::do_apply_event(rpl_group_info *rgi)
 
       error= do_exec_row(rgi);
 
+      DBUG_EXECUTE_IF("simulate_long_running_row_event",
+                      { my_sleep(1000000); });
```

Comment:
avoid debug tests whenever possible. we have about 50 non-debug builders and 1 (one!) debug builder. You should try to avoid reducing the coverage of your tests by a factor of 50, `have_debug.inc` is a last resort measure.

Here you likely don't need it, you can run `LOCK TABLES` on the slave and make any row event to wait until it times out.

Reply:
Locking-based was tried. Just two points. 
LOCK TABLES does not let to error-stop in between of record operations (because this lock is acquired at the statement locking time).
I tried yesterday to refine to the record-level, but was not fully content with that `THD::main_da::error` would be set to `1205` (which is treated exceptionally [as a retriable error] in many slave branches). Let me look into that for the last time..

Reply:
well, as I said, you can make the test run on all builders — you must do it. and in this case you can. If you want to stop it between rows, use InnoDB row locks to lock an individual row